File: frontend/src/types/v2/template-metadata.ts
/**
 * Extended Template Metadata Types (Vitec Parity)
 */

/**
 * Template channel types
 */
export type TemplateChannel = 'pdf' | 'email' | 'sms' | 'pdf_email';

/**
 * Template type (Vitec classification)
 */
export type TemplateType = 'Objekt/Kontakt' | 'System';

/**
 * Receiver type
 */
export type ReceiverType = 'Egne/kundetilpasset' | 'Systemstandard';

/**
 * Receiver roles (dynamic list from Vitec/Proaktiv)
 */
export type Receiver = string;

export interface TemplateTag {
  id: string;
  name: string;
  color: string;
}

/**
 * Transaction phases
 */
export type TransactionPhase = 
  | 'Oppdrag' 
  | 'Markedsføring' 
  | 'Visning' 
  | 'Budrunde' 
  | 'Kontrakt' 
  | 'Oppgjør';

/**
 * Ownership types
 */
export type OwnershipType = 'Bolig' | 'Aksje' | 'Tomt' | 'Næring' | 'Hytte';

/**
 * Extended template with Vitec metadata fields
 */
export interface TemplateWithMetadata {
  id: string;
  title: string;
  description: string | null;
  file_name: string;
  file_type: string;
  file_size_bytes: number;
  status: 'draft' | 'published' | 'archived';
  version: number;
  preview_url: string | null;
  created_at: string | null;
  updated_at: string | null;
  attachments?: string[];
  tags?: TemplateTag[];
  
  // V2 Vitec Metadata Fields
  preview_thumbnail_url: string | null;
  channel: TemplateChannel;
  template_type: TemplateType | null;
  receiver_type: ReceiverType | null;
  receiver: Receiver | null;
  extra_receivers: string[];
  phases: TransactionPhase[];
  assignment_types: string[];
  ownership_types: OwnershipType[];
  departments: string[];
  email_subject: string | null;
  header_template_id: string | null;
  footer_template_id: string | null;
  margin_top: number;
  margin_bottom: number;
  margin_left: number;
  margin_right: number;
}

/**
 * Template metadata update payload
 */
export interface TemplateMetadataUpdate {
  channel?: TemplateChannel;
  template_type?: TemplateType;
  receiver_type?: ReceiverType;
  receiver?: Receiver;
  extra_receivers?: string[];
  phases?: TransactionPhase[];
  assignment_types?: string[];
  ownership_types?: OwnershipType[];
  departments?: string[];
  email_subject?: string;
  header_template_id?: string | null;
  footer_template_id?: string | null;
  margin_top?: number;
  margin_bottom?: number;
  margin_left?: number;
  margin_right?: number;
  preview_thumbnail_url?: string;
}

/**
 * Template analysis result
 */
export interface TemplateAnalysisResult {
  template_id: string;
  merge_fields_found: string[];
  conditions_found: string[];
  loops_found: string[];
  unknown_fields: string[];
  analysis_timestamp: string;
}

File: backend/app/schemas/template_metadata.py
"""
Pydantic schemas for extended Template Metadata (Vitec parity).
"""

from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class TemplateMetadataUpdate(BaseModel):
    """
    Schema for updating template Vitec metadata fields.
    All fields are optional for partial updates.
    """

    # Channel and Type
    channel: Literal["pdf", "email", "sms", "pdf_email"] | None = Field(None, description="Template channel type")
    template_type: Literal["Objekt/Kontakt", "System"] | None = Field(None, description="Template type")

    # Receiver
    receiver_type: Literal["Egne/kundetilpasset", "Systemstandard"] | None = Field(None, description="Receiver type")
    receiver: (
        Literal[
            "Selger",
            "Kjøper",
            "Megler",
            "Bank",
            "Forretningsfører",
            "Visningsdeltager",
            "Budgiver",
            "Interessent",
            "Interessent - Match",
            "Utleier",
            "Leietaker",
            "Borettslag",
            "Bank - selger",
            "Bank - kjøper",
            "Ikke interessert",
            "Hjemmelshaver",
            "Panthaver",
            "Saksøkt",
            "Kreditor",
            "Prosessfullmektig",
            "Saksøker",
            "Opprinnelig kjøper",
            "Fremleier",
            "Fremleietaker",
            "Takstmann",
            "Sameie",
            "Aksjeselskap",
            "Kommune",
            "Interessent - budoppfølging",
            "Interessent - autoprospekt",
            "Fester",
            "Fakturamottaker",
            "Hjemmelsutsteder",
            "Fullmektig",
            "Grunneier",
            "Tidligere eier/avdød",
            "Arving",
            "Selgers ektefelle",
            "Kontodisponent selger",
            "Kontodisponent kjøper",
            "Kartverket",
            "Advokat",
            "Fotograf",
            "Samarbeidspartner",
            "Annen",
            "Bortefester",
            "Tinglysning",
            "Forening",
            "Forsikringsselskap",
            "Kjoper",
            "Forretningsforer",
        ]
        | None
    ) = Field(None, description="Primary receiver")
    extra_receivers: list[str] | None = Field(None, description="Additional receivers")

    # Filtering/Categorization
    phases: list[Literal["Oppdrag", "Markedsføring", "Visning", "Budrunde", "Kontrakt", "Oppgjør"]] | None = Field(
        None, description="Applicable phases"
    )
    assignment_types: list[str] | None = Field(None, description="Assignment types")
    ownership_types: list[Literal["Bolig", "Aksje", "Tomt", "Næring", "Hytte"]] | None = Field(
        None, description="Ownership types"
    )
    departments: list[str] | None = Field(None, description="Departments")

    # Email
    email_subject: str | None = Field(None, max_length=500, description="Email subject line (can include merge fields)")

    # Layout
    header_template_id: UUID | None = Field(None, description="Header layout partial ID")
    footer_template_id: UUID | None = Field(None, description="Footer layout partial ID")

    # Margins (in cm)
    margin_top: Decimal | None = Field(None, ge=0, le=10, description="Top margin in cm")
    margin_bottom: Decimal | None = Field(None, ge=0, le=10, description="Bottom margin in cm")
    margin_left: Decimal | None = Field(None, ge=0, le=10, description="Left margin in cm")
    margin_right: Decimal | None = Field(None, ge=0, le=10, description="Right margin in cm")

    # Thumbnail
    preview_thumbnail_url: str | None = Field(None, description="Preview thumbnail URL")


class TemplateAnalysisResult(BaseModel):
    """Schema for template analysis result."""

    template_id: UUID
    merge_fields_found: list[str]
    conditions_found: list[str]
    loops_found: list[str]
    unknown_fields: list[str]
    analysis_timestamp: str
